AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/PreferencesHolder.java

```diff
@@ -0,0 +1,163 @@
+		if (addPreferencesFromResourceBlock == null) {
+			setAddPreferencesFromResourceBlock();
+		}
+		return addPreferencesFromResourceBlock;
+	}
+
+	private void setAddPreferencesFromResourceBlock() {
+		JMethod method = getGeneratedClass().method(PUBLIC, codeModel().VOID, "addPreferencesFromResource");
+		method.annotate(Override.class);
+		JVar preferencesResIdParam = method.param(int.class, "preferencesResId");
+		method.body().invoke(JExpr._super(), "addPreferencesFromResource").arg(preferencesResIdParam);
+		addPreferencesFromResourceBlock = method.body();
+	}
+
+	private JInvocation findPreferenceByKey(JFieldRef idRef) {
+		JInvocation getString = invoke(_this(), "getString").arg(idRef);
+		JInvocation findPreferenceByKey = invoke(_this(), "findPreference");
+		return findPreferenceByKey.arg(getString);
+	}
+
+	@Override
+	public void assignFindPreferenceByKey(JFieldRef idRef, JClass preferenceClass, JFieldRef fieldRef) {
+		String idRefString = codeModelHelper.getIdStringFromIdFieldRef(idRef);
+		FoundPreferenceHolder foundViewHolder = (FoundPreferenceHolder) holder.foundHolders.get(idRefString);
+
+		JBlock block = getAddPreferencesFromResourceBlock();
+		JExpression assignExpression;
+
+		if (foundViewHolder != null) {
+			assignExpression = foundViewHolder.getOrCastRef(preferenceClass);
+		} else {
+			assignExpression = findPreferenceByKey(idRef);
+			if (preferenceClass != null && preferenceClass != classes().PREFERENCE) {
+				assignExpression = cast(preferenceClass, assignExpression);
+			}
+			holder.foundHolders.put(idRefString, new FoundPreferenceHolder(this, preferenceClass, fieldRef, block));
+		}
+
+		block.assign(fieldRef, assignExpression);
+	}
+
+	@Override
+	public FoundPreferenceHolder getFoundPreferenceHolder(JFieldRef idRef, JClass preferenceClass) {
+		String idRefString = codeModelHelper.getIdStringFromIdFieldRef(idRef);
+		FoundPreferenceHolder foundPreferenceHolder = (FoundPreferenceHolder) holder.foundHolders.get(idRefString);
+		if (foundPreferenceHolder == null) {
+			foundPreferenceHolder = createFoundPreferenceAndIfNotNullBlock(idRef, preferenceClass);
+			holder.foundHolders.put(idRefString, foundPreferenceHolder);
+		}
+		return foundPreferenceHolder;
+	}
+
+	private FoundPreferenceHolder createFoundPreferenceAndIfNotNullBlock(JFieldRef idRef, JClass preferenceClass) {
+		JExpression findPreferenceExpression = findPreferenceByKey(idRef);
+		JBlock block = getAddPreferencesFromResourceBlock().block();
+
+		if (preferenceClass == null) {
+			preferenceClass = classes().PREFERENCE;
+		} else if (preferenceClass != classes().PREFERENCE) {
+			findPreferenceExpression = cast(preferenceClass, findPreferenceExpression);
+		}
+
+		JVar preference = block.decl(preferenceClass, "preference", findPreferenceExpression);
+		return new FoundPreferenceHolder(this, preferenceClass, preference, block);
+	}
+
+	@Override
+	public JBlock getPreferenceScreenInitializationBlock() {
+		// not used
```

Comment:
@yDelouis the returned block is different in `EFragmentHolder` and `EActivityHolder`. It is easier to just return the appropriate block in those classes and not delegating to this. If you think it is not clean, we can create a subclass like `PreferenceFragmentHolder` (like `PreferenceActivityHolder`), and return the init blocks there.